iot/mqttsnippets/gateway_send_data_test.go

```diff
@@ -0,0 +1,284 @@
+	client, err := getClient()
+	if err != nil {
+		return nil, err
+	}
+
+	parent := fmt.Sprintf("projects/%s/locations/%s/registries/%s", projectID, region, registryID)
+	bindRequest := &cloudiot.BindDeviceToGatewayRequest{
+		DeviceId:  deviceID,
+		GatewayId: gatewayID,
+	}
+
+	response, err := client.Projects.Locations.Registries.BindDeviceToGateway(parent, bindRequest).Do()
+
+	if err != nil {
+		return nil, fmt.Errorf("BindDeviceToGateway: %v", err)
+	}
+
+	if response.HTTPStatusCode/100 != 2 {
+		return nil, fmt.Errorf("BindDeviceToGateway: HTTP status code not 2xx\n %v", response)
+	}
+
+	fmt.Fprintf(w, "Bound %s to %s", deviceID, gatewayID)
+
+	return response, nil
+}
+
+// unbindDeviceFromGateway unbinds a bound device from a gateway.
+func unbindDeviceFromGateway(w io.Writer, projectID string, region string, registryID string, gatewayID string, deviceID string) (*cloudiot.UnbindDeviceFromGatewayResponse, error) {
+	client, err := getClient()
+	if err != nil {
+		return nil, err
+	}
+
+	parent := fmt.Sprintf("projects/%s/locations/%s/registries/%s", projectID, region, registryID)
+	unbindRequest := &cloudiot.UnbindDeviceFromGatewayRequest{
+		DeviceId:  deviceID,
+		GatewayId: gatewayID,
+	}
+
+	response, err := client.Projects.Locations.Registries.UnbindDeviceFromGateway(parent, unbindRequest).Do()
+
+	if err != nil {
+		return nil, fmt.Errorf("UnbindDeviceFromGateway error: %v", err)
+	}
+
+	if response.HTTPStatusCode/100 != 2 {
+		return nil, fmt.Errorf("UnbindDeviceFromGateway: HTTP status code not 2xx\n %v", response)
+	}
+
+	fmt.Fprintf(w, "Unbound %s from %s", deviceID, gatewayID)
+
+	return response, nil
+}
+
+func TestSendDataFromBoundDevice(t *testing.T) {
+	projectID := testutil.SystemTest(t).ProjectID
+
+	registryID := "golang-iot-test-registry"
+	gatewayID := createIDForTest("gateway")
+	deviceID := createIDForTest("device")
+
+	testutil.Retry(t, 1, 10*time.Second, func(r *testutil.R) {
+		if _, err := createGateway(ioutil.Discard, projectID, region, registryID, gatewayID, "ASSOCIATION_ONLY", pubKeyRSA); err != nil {
+			r.Errorf("Could not create gateway: %v\n", err)
+			return
+		}
+
+		if _, err := createDevice(ioutil.Discard, projectID, region, registryID, deviceID, "RSA_X509_PEM", pubKeyRSA); err != nil {
+			r.Errorf("Could not create device: %v\n", err)
+			return
+		}
+
+		if _, err := bindDeviceToGateway(ioutil.Discard, projectID, region, registryID, gatewayID, deviceID); err != nil {
+			r.Errorf("Could not bind device to gateway: %v\n", err)
+			return
+		}
+
+		buf := new(bytes.Buffer)
+		sendDataFromBoundDevice(buf, projectID, region, registryID, gatewayID, deviceID, privateKeyRSA, "RS256", 2, "test")
+
+		want := fmt.Sprintf("Publishing message: %s", "test")
+		if got := buf.String(); !strings.Contains(got, want) {
+			r.Errorf("SendDataFromBoundDevice got %s, want substring %q", got, want)
+		}
+	})
+
+	// cleanup
+	unbindDeviceFromGateway(ioutil.Discard, projectID, region, registryID, gatewayID, deviceID)
```

Comment:
Might be good to have a retry for these, too.

Reply:
Done.